sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py

```diff
@@ -110,6 +110,11 @@ def __init__(
             endpoint=endpoint, credential=credential, credential_scopes=defaultScope, **kwargs
         )
 
+    def __enter__(self) -> "ContainerRegistryClient":
+        self._client.__enter__()
+        self._auth_policy.__enter__()
```

Comment:
Do we need to explicitly enter auth_policy?

Isn't it part of the pipeline?

Reply:
Why did we move this from the base client to the main client? If it's because of typing / circular import with the return type, you can put it under a TYPE_CHECKING check

Reply:
> Why did we move this from the base client to the main client? If it's because of typing / circular import with the return type, you can put it under a TYPE_CHECKING check

It's because mypy think the client is in base client type if we don't move (the return type was base client type), and will have an error in every function call using this client. For example: https://dev.azure.com/azure-sdk/public/_build/results?buildId=3119388&view=logs&j=b70e5e73-bbb6-5567-0939-8415943fadb9&t=e3f8f308-6e5f-58aa-3a2a-fe425958de32&l=86

Reply:
> Do we need to explicitly enter auth_policy?
> 
> Isn't it part of the pipeline?

We need. You added it to resolve this customer issue https://github.com/Azure/azure-sdk-for-python/issues/20000 in this PR: https://github.com/Azure/azure-sdk-for-python/pull/20047/files#diff-b5f4ab365871cf40ef14a049f86d31a30a8994eead16610d6925b48da0539839 :laughing:

Reply:
> > Why did we move this from the base client to the main client? If it's because of typing / circular import with the return type, you can put it under a TYPE_CHECKING check
> 
> It's because mypy think the client is in base client type if we don't move (the return type was base client type), and will have an error in every function call using this client. For example: https://dev.azure.com/azure-sdk/public/_build/results?buildId=3119388&view=logs&j=b70e5e73-bbb6-5567-0939-8415943fadb9&t=e3f8f308-6e5f-58aa-3a2a-fe425958de32&l=86

Found a way to resolve the odd mypy issue, have moved it back.